### PR TITLE
New: Added button trait to core [ED-4597]

### PR DESCRIPTION
**Pull request by nisan250**

### CODE REVIEW

File: includes/widgets/traits/button.php [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+use Elementor\Core\Kits\Documents\Tabs\Global_Colors;
+use Elementor\Core\Kits\Documents\Tabs\Global_Typography;
+use Elementor\Group_Control_Background;
+use Elementor\Group_Control_Border;
+use Elementor\Group_Control_Box_Shadow;
+use Elementor\Group_Control_Text_Shadow;
+use Elementor\Group_Control_Typography;
+use Elementor\Icons_Manager;
+use Elementor\Widget_Base;
+
```
Comment (matipojo) [inline includes/widgets/traits/button.php:13]: Exit if not ABSPATH

File: includes/widgets/traits/button.php [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+	 *
+	 * @since 3.4.0
+	 * @access public
+	 * @static
+	 *
+	 * @return array An array containing button sizes.
+	 */
+	public static function get_button_sizes() {
+		return [
+			'xs' => __( 'Extra Small', 'elementor-pro' ),
```
Comment (matipojo) [inline includes/widgets/traits/button.php:28]:
```suggestion
			'xs' => __( 'Extra Small', 'elementor' ),
```

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]
Use the trait in the button widget

File: includes/widgets/traits/button.php [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+	}
+
+	protected function register_button_content_controls( $args = [] ) {
+		$default_args = [
+			'section_condition' => [],
+			'button_text' => __( 'Click here', 'elementor-pro' ),
+			'control_label_name' => __( 'Text', 'elementor-pro' ),
+			'prefix_class' => 'elementor%s-align-',
+			'alignment_default' => '',
+			'exclude_inline_options' => [],
```
Comment (matipojo) [inline includes/widgets/traits/button.php:43]:
```suggestion
			'icon_exclude_inline_options' => [],
```

File: includes/widgets/traits/button.php [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+			'sm' => __( 'Small', 'elementor-pro' ),
+			'md' => __( 'Medium', 'elementor-pro' ),
+			'lg' => __( 'Large', 'elementor-pro' ),
+			'xl' => __( 'Extra Large', 'elementor-pro' ),
+		];
+	}
+
+	protected function register_button_content_controls( $args = [] ) {
+		$default_args = [
+			'section_condition' => [],
```
Comment (matipojo) [inline includes/widgets/traits/button.php:38]: Add docs for each option (PHPDoc for the $args)

**Review by matipojo — CHANGES_REQUESTED** [review]

File: includes/widgets/traits/button.php (hunk 1 above)
Comment (nisan250) [reply includes/widgets/traits/button.php:13]: Done

File: includes/widgets/traits/button.php (hunk 2 above)
Comment (nisan250) [reply includes/widgets/traits/button.php:28]: Done

File: includes/widgets/traits/button.php (hunk 3 above)
Comment (nisan250) [reply includes/widgets/traits/button.php:43]: Done

File: includes/widgets/traits/button.php (hunk 4 above)
Comment (nisan250) [reply includes/widgets/traits/button.php]: Done

File: includes/widgets/button.php [hunk 5]
```diff
@@ -113,163 +85,7 @@ protected function register_controls() {
...
-					/* translators: %1$s Code open tag, %2$s: Code close tag. */
-					esc_html__( 'Please make sure the ID is unique and not used elsewhere on the page this form is displayed. This field allows %1$sA-z 0-9%2$s & underscore chars without spaces.', 'elementor' ),
-					'<code>',
-					'</code>'
-				),
-				'separator' => 'before',
-
-			]
-		);
+		$this->register_button_content_controls( [] );
```
Comment (Nevoss) [inline includes/widgets/button.php:88]:
```suggestion
		$this->register_button_content_controls();
```

File: includes/widgets/traits/button-trait.php [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+				'default' => '',
+				'options' => [
+					'' => esc_html__( 'Default', 'elementor' ),
+					'info' => esc_html__( 'Info', 'elementor' ),
+					'success' => esc_html__( 'Success', 'elementor' ),
+					'warning' => esc_html__( 'Warning', 'elementor' ),
+					'danger' => esc_html__( 'Danger', 'elementor' ),
+				],
+				'prefix_class' => 'elementor-button-',
+				'condition' => $args['section_condition'],
```
Comment (Nevoss) [inline includes/widgets/traits/button-trait.php:81]: make sure this is visible if the condition are empty

File: includes/widgets/traits/button-trait.php [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+	 *     @type string $control_label_name  Name for the label of the text control.
+	 *     @type string $prefix_class  Prefix class name for the button alignment control.
+	 *     @type string $alignment_default  Default alignment for the button.
+	 *     @type array $icon_exclude_inline_options  Set of icon types to exclude from icon controls.
+	 * }
+	 */
+	protected function register_button_content_controls( $args = [] ) {
+		$default_args = [
+			'section_condition' => [],
+			'button_text' => esc_html__( 'Click here', 'elementor' ),
```
Comment (Nevoss) [inline includes/widgets/traits/button-trait.php:58]:
```suggestion
			'button_default_text' => esc_html__( 'Click here', 'elementor' ),
```
only if possible

File: includes/widgets/traits/button-trait.php [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+			'control_label_name' => esc_html__( 'Text', 'elementor' ),
```
Comment (Nevoss) [inline includes/widgets/traits/button-trait.php:59]: Which control?
```suggestion
			'text_control_label' => esc_html__( 'Text', 'elementor' ),
```

File: includes/widgets/traits/button-trait.php [hunk 9]
```diff
@@ -0,0 +1,524 @@
...
+			'prefix_class' => 'elementor%s-align-',
```
Comment (Nevoss) [inline includes/widgets/traits/button-trait.php:60]:
```suggestion
			'alignment_control_prefix_class' => 'elementor%s-align-',
```

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: includes/widgets/button.php (hunk 5 above)
Comment (nisan250) [reply includes/widgets/button.php:88]: Done

File: includes/widgets/traits/button-trait.php (hunk 6 above)
Comment (nisan250) [reply includes/widgets/traits/button-trait.php:81]: It is visible

File: includes/widgets/traits/button-trait.php (hunk 7 above)
Comment (nisan250) [reply includes/widgets/traits/button-trait.php:58]: Done

File: includes/widgets/traits/button-trait.php (hunk 8 above)
Comment (nisan250) [reply includes/widgets/traits/button-trait.php:59]: Done

File: includes/widgets/traits/button-trait.php (hunk 9 above)
Comment (nisan250) [reply includes/widgets/traits/button-trait.php:60]: Done

File: includes/widgets/traits/button-trait.php [hunk 10]
```diff
@@ -0,0 +1,524 @@
...
+		if ( ! empty( $settings['size'] ) ) {
+			$instance->add_render_attribute( 'button', 'class', 'elementor-size-' . $settings['size'] );
+		}
+
+		if ( $settings['hover_animation'] ) {
+			$instance->add_render_attribute( 'button', 'class', 'elementor-animation-' . $settings['hover_animation'] );
+		}
+		?>
+		<div <?php echo $instance->get_render_attribute_string( 'wrapper' ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>>
+			<a <?php echo $instance->get_render_attribute_string( 'button' ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>>
```
Comment (matipojo) [inline includes/widgets/traits/button-trait.php:458]: Use `print_render_attribute_string` here and next

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]
Please pull from 3.5.0

**Review by matipojo — APPROVED** [review]

File: includes/widgets/traits/button-trait.php [hunk 11]
```diff
@@ -0,0 +1,559 @@
...
+		if ( ! empty( $settings['size'] ) ) {
+			$instance->add_render_attribute( 'button', 'class', 'elementor-size-' . $settings['size'] );
+		}
+
+		if ( ! empty( $settings['hover_animation'] ) ) {
+			$instance->add_render_attribute( 'button', 'class', 'elementor-animation-' . $settings['hover_animation'] );
+		}
+		?>
+		<div <?php $instance->print_render_attribute_string( 'wrapper' ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>>
+			<a <?php $instance->print_render_attribute_string( 'button' ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>>
```
Comment (matipojo) [inline includes/widgets/traits/button-trait.php:458]:
```suggestion
		<div <?php $instance->print_render_attribute_string( 'wrapper' ); ?>>
			<a <?php $instance->print_render_attribute_string( 'button' ); ?>>
```

File: includes/widgets/traits/button-trait.php [hunk 12]
```diff
@@ -0,0 +1,559 @@
...
+			<?php if ( ! empty( $settings['icon'] ) || ! empty( $settings['selected_icon']['value'] ) ) : ?>
+				<span <?php $instance->print_render_attribute_string( 'icon-align' ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>>
+				<?php if ( $is_new || $migrated ) :
+					Icons_Manager::render_icon( $settings['selected_icon'], [ 'aria-hidden' => 'true' ] );
+				else : ?>
+					<i class="<?php echo esc_attr( $settings['icon'] ); ?>" aria-hidden="true"></i>
+				<?php endif; ?>
+			</span>
+			<?php endif; ?>
+			<span <?php $instance->print_render_attribute_string( 'text' ); ?>><?php echo $settings['text']; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?></span>
```
Comment (matipojo) [inline includes/widgets/traits/button-trait.php:551]:
```suggestion
		<span <?php $instance->print_render_attribute_string( 'content-wrapper' ); ?>>
			<?php if ( ! empty( $settings['icon'] ) || ! empty( $settings['selected_icon']['value'] ) ) : ?>
				<span <?php $instance->print_render_attribute_string( 'icon-align' ); ?>>
				<?php if ( $is_new || $migrated ) :
					Icons_Manager::render_icon( $settings['selected_icon'], [ 'aria-hidden' => 'true' ] );
				else : ?>
					<i class="<?php echo esc_attr( $settings['icon'] ); ?>" aria-hidden="true"></i>
				<?php endif; ?>
			</span>
			<?php endif; ?>
			<span <?php $instance->print_render_attribute_string( 'text' ); ?>><?php $this->print_unescaped_setting( 'text' ); ?></span>
```

**Review by matipojo — CHANGES_REQUESTED** [review]

**Pull request merged**